File: js/src/core/page-trigger-manager.js
/**
 * PageTriggerManager - Manages stoppable page triggers with proper lifecycle
 *
 * Key guarantees:
 * 1. Only one action runs at a time
 * 2. Action is fully stopped before page loading starts
 * 3. Action can gracefully cleanup on stop
 * 4. All commander operations throw ActionStoppedError when stopped
 *
 * Terminology:
 * - Trigger: A condition + action pair that fires when condition is met
 * - Condition: A function that returns true/false to determine if trigger should fire
 * - Action: The async function that runs when condition is true
 */

/**
 * Error thrown when action is stopped (navigation detected)
 * Actions can catch this to do cleanup, but should re-throw or return
 */
export class ActionStoppedError extends Error {
  constructor(message = 'Action stopped due to navigation') {
    super(message);
    this.name = 'ActionStoppedError';
    this.isActionStopped = true;
  }
}

/**
 * Check if error is an ActionStoppedError
 * @param {Error} error - Error to check
 * @returns {boolean}
 */
export function isActionStoppedError(error) {
  return (
    error &&
    (error.isActionStopped === true || error.name === 'ActionStoppedError')
  );
}

/**
 * Create a URL condition matcher (similar to express router patterns)
 *
 * @param {string|RegExp|Function} pattern - URL pattern to match
 *   - String: Exact match or pattern with :param placeholders (like express)
 *   - RegExp: Regular expression to test against URL
 *   - Function: Custom function (url, ctx) => boolean
 * @returns {Function} - Condition function (ctx) => boolean
 *
 * @example
 * // Exact match
 * makeUrlCondition('https://example.com/page')
 *
 * // Pattern with parameters (express-style)
 * makeUrlCondition('/vacancy/:id')
 * makeUrlCondition('https://hh.ru/vacancy/:vacancyId')
 *
 * // Contains substring
 * makeUrlCondition('*checkout*')  // matches any URL containing 'checkout'
 *
 * // RegExp
 * makeUrlCondition(/\/product\/\d+/)
 *
 * // Custom function
 * makeUrlCondition((url, ctx) => url.includes('/admin') && url.includes('edit'))
 */
export function makeUrlCondition(pattern) {
  // If already a function, wrap it to receive context
  if (typeof pattern === 'function') {
    return (ctx) => pattern(ctx.url, ctx);
  }

  // RegExp pattern
  if (pattern instanceof RegExp) {
    return (ctx) => pattern.test(ctx.url);
  }

  // String pattern
  if (typeof pattern === 'string') {
    // Wildcard pattern: *substring* means "contains"
    if (pattern.startsWith('*') && pattern.endsWith('*')) {
      const substring = pattern.slice(1, -1);
      return (ctx) => ctx.url.includes(substring);
    }

    // Starts with wildcard: *suffix means "ends with"
    if (pattern.startsWith('*')) {
      const suffix = pattern.slice(1);
      return (ctx) => ctx.url.endsWith(suffix);
    }

    // Ends with wildcard: prefix* means "starts with"
    if (pattern.endsWith('*')) {
      const prefix = pattern.slice(0, -1);
      return (ctx) => ctx.url.startsWith(prefix);
    }

    // Express-style pattern with :params
    if (pattern.includes(':')) {
      // Convert express pattern to regex
      // /vacancy/:id -> /vacancy/([^/]+)
      // /user/:userId/profile -> /user/([^/]+)/profile
      const regexPattern = pattern
        .replace(/[.*+?^${}()|[\]\\]/g, '\\$&') // Escape special chars first
        .replace(/\\:/g, ':') // Unescape colons we just escaped
        .replace(/:([a-zA-Z_][a-zA-Z0-9_]*)/g, '([^/&?#]+)'); // Replace :param with capture group

      const regex = new RegExp(regexPattern);
      return (ctx) => regex.test(ctx.url);
    }

    // Simple substring match (no wildcards, no params)
    // If it looks like a full URL, do exact match
    if (pattern.startsWith('http://') || pattern.startsWith('https://')) {
      return (ctx) =>
        ctx.url === pattern ||
        ctx.url.startsWith(`${pattern}?`) ||
        ctx.url.startsWith(`${pattern}#`);
    }

    // Otherwise, treat as path pattern - match if URL contains this path
    return (ctx) => ctx.url.includes(pattern);
  }

  throw new Error(
    `Invalid URL pattern type: ${typeof pattern}. Expected string, RegExp, or function.`
  );
}

/**
 * Combine multiple conditions with AND logic
 * @param {...Function} conditions - Condition functions to combine
 * @returns {Function} - Combined condition function
 */
export function allConditions(...conditions) {
  return (ctx) => conditions.every((cond) => cond(ctx));
}

/**
 * Combine multiple conditions with OR logic
 * @param {...Function} conditions - Condition functions to combine
 * @returns {Function} - Combined condition function
 */
export function anyCondition(...conditions) {
  return (ctx) => conditions.some((cond) => cond(ctx));
}

/**
 * Negate a condition
 * @param {Function} condition - Condition function to negate
 * @returns {Function} - Negated condition function
 */
export function notCondition(condition) {
  return (ctx) => !condition(ctx);
}

/**
 * Create a PageTriggerManager instance
 * @param {Object} options - Configuration options
 * @param {Object} options.navigationManager - NavigationManager instance
 * @param {Function} options.log - Logger instance
 * @returns {Object} - PageTriggerManager API
 */
export function createPageTriggerManager(options = {}) {
  const { navigationManager, log } = options;

  if (!navigationManager) {
    throw new Error('navigationManager is required');
  }

  // Registered triggers
  const triggers = [];

  // Current action state
  let currentTrigger = null;
  let currentAbortController = null;
  let actionPromise = null;
  let actionStopPromise = null;
  let actionStopResolve = null;
  let isActionRunning = false;
  let isStopping = false;

  /**
   * Register a page trigger
   * @param {Object} config - Trigger configuration
   * @param {Function} config.condition - Function (ctx) => boolean, returns true if trigger should fire
   * @param {Function} config.action - Async function (ctx) => void, the action to run
   * @param {string} config.name - Trigger name for debugging
   * @param {number} config.priority - Priority (higher runs first if multiple match), default 0
   * @returns {Function} - Unregister function
   */
  function pageTrigger(config) {
    const { condition, action, name = 'unnamed', priority = 0 } = config;

    if (typeof condition !== 'function') {
      throw new Error('condition must be a function');
    }
    if (typeof action !== 'function') {
      throw new Error('action must be a function');
    }

    const triggerConfig = {
      condition,
      action,
      name,
      priority,
    };

    triggers.push(triggerConfig);

    // Sort by priority (descending)
    triggers.sort((a, b) => b.priority - a.priority);

    log.debug(
      () => `📋 Registered page trigger: "${name}" (priority: ${priority})`
    );

    // Return unregister function
    return () => {
      const index = triggers.indexOf(triggerConfig);
      if (index !== -1) {
        triggers.splice(index, 1);
        log.debug(() => `📋 Unregistered page trigger: "${name}"`);
      }
    };
  }

  /**
   * Find matching trigger for context
   * @param {Object} ctx - Context with url and other properties
   * @returns {Object|null} - Matching trigger config or null
   */
  function findMatchingTrigger(ctx) {
    for (const config of triggers) {
      try {
        if (config.condition(ctx)) {
          return config;
        }
      } catch (e) {
        log.debug(
          () => `⚠️  Error in condition for "${config.name}": ${e.message}`
        );
      }
    }
    return null;
  }

  /**
   * Stop current action and wait for it to finish
   * @returns {Promise<void>}
   */
  async function stopCurrentAction() {
    if (!isActionRunning) {
      return;
    }

    if (isStopping) {
      // Already stopping, wait for it
      if (actionStopPromise) {
        await actionStopPromise;
      }
      return;
    }

    isStopping = true;
    log.debug(() => `🛑 Stopping action "${currentTrigger?.name}"...`);

    // Create promise that resolves when action actually stops
    actionStopPromise = new Promise((resolve) => {
      actionStopResolve = resolve;
    });

    // Abort the action
    if (currentAbortController) {
      currentAbortController.abort();
    }

    // Wait for action to finish (with timeout)
    const timeoutMs = 10000; // 10 second max wait
    const timeoutPromise = new Promise((resolve) => {
      setTimeout(() => {
        log.debug(
          () =>
            `⚠️  Action "${currentTrigger?.name}" did not stop gracefully within ${timeoutMs}ms`
        );
        resolve();
      }, timeoutMs);
    });

    await Promise.race([actionPromise, timeoutPromise]);

    // Cleanup
    isActionRunning = false;
    isStopping = false;
    currentTrigger = null;
    currentAbortController = null;
    actionPromise = null;

    // Resolve the stop promise
    if (actionStopResolve) {
      actionStopResolve();
      actionStopResolve = null;
      actionStopPromise = null;
    }

    log.debug(() => '✅ Action stopped');
  }

  /**
   * Start action for URL
   * @param {string} url - URL to start action for
   * @param {Object} commander - BrowserCommander instance
   */
  async function startAction(url, commander) {
    // Create context for condition checking
    const conditionCtx = {
      url,
      commander,
    };

    // Find matching trigger
    const matchingTrigger = findMatchingTrigger(conditionCtx);
    if (!matchingTrigger) {
      log.debug(() => `📋 No trigger registered for: ${url}`);
      return;
    }

    log.debug(() => `🚀 Starting action "${matchingTrigger.name}" for: ${url}`);

    // Setup abort controller
    currentAbortController = new AbortController();
    currentTrigger = matchingTrigger;
    isActionRunning = true;

    // Create action context
    const context = createActionContext({
      url,
      abortSignal: currentAbortController.signal,
      commander,
      triggerName: matchingTrigger.name,
    });

    // Run action
    actionPromise = (async () => {
      try {
        await matchingTrigger.action(context);
        log.debug(
          () => `✅ Action "${matchingTrigger.name}" completed normally`
        );
      } catch (error) {
        if (isActionStoppedError(error)) {
          log.debug(
            () =>
              `🛑 Action "${matchingTrigger.name}" stopped (caught ActionStoppedError)`
          );
        } else if (error.name === 'AbortError') {
          log.debug(() => `🛑 Action "${matchingTrigger.name}" aborted`);
        } else {
          log.debug(
            () => `❌ Action "${matchingTrigger.name}" error: ${error.message}`
          );
          console.error(`Action "${matchingTrigger.name}" error:`, error);
        }
      } finally {
        // Only clear if this is still the current trigger
        if (currentTrigger === matchingTrigger) {
          isActionRunning = false;
          currentTrigger = null;
          currentAbortController = null;
        }
      }
    })();
  }

  /**
   * Create action context with abort-aware commander wrapper
   * @param {Object} options
   * @returns {Object} - Action context
   */
  function createActionContext(options) {
    const { url, abortSignal, commander, triggerName } = options;

    /**
     * Check if stopped and throw if so
     */
    function checkStopped() {
      if (abortSignal.aborted) {
        throw new ActionStoppedError(`Action "${triggerName}" stopped`);
      }
    }

    /**
     * Wrap async function to check abort before and after
     */
    function wrapAsync(fn) {
      return async (...args) => {
        checkStopped();
        const result = await fn(...args);
        checkStopped();
        return result;
      };
    }

    /**
     * Create abort-aware loop helper
     * Use this instead of for/while loops for stoppability
     */
    async function forEach(items, callback) {
      for (let i = 0; i < items.length; i++) {
        checkStopped();
        await callback(items[i], i, items);
      }
    }

    /**
     * Wait with abort support
     */
    async function wait(ms) {
      checkStopped();
      await new Promise((resolve, reject) => {
        const timeout = setTimeout(resolve, ms);
        abortSignal.addEventListener(
          'abort',
          () => {
            clearTimeout(timeout);
            reject(new ActionStoppedError());
          },
          { once: true }
        );
      });
    }

    /**
     * Register cleanup callback (called when action stops)
     */
    const cleanupCallbacks = [];
    function onCleanup(callback) {
      cleanupCallbacks.push(callback);
      abortSignal.addEventListener(
        'abort',
        async () => {
          try {
            await callback();
          } catch (e) {
            log.debug(() => `⚠️  Cleanup error: ${e.message}`);
          }
        },
        { once: true }
      );
    }

    // Wrap all commander methods to be abort-aware
    const wrappedCommander = {};
    for (const [key, value] of Object.entries(commander)) {
      if (
        typeof value === 'function' &&
        key !== 'destroy' &&
        key !== 'pageTrigger'
      ) {
        wrappedCommander[key] = wrapAsync(value);
      } else {
        wrappedCommander[key] = value;
      }
    }

    return {
      // URL this action is running for
      url,

      // Abort signal - use with fetch() or custom abort logic
      abortSignal,

      // Check if action should stop
      isStopped: () => abortSignal.aborted,

      // Throw if stopped - call this in loops
      checkStopped,

      // Abort-aware iteration helper
      forEach,

      // Abort-aware wait
      wait,

      // Register cleanup callback
      onCleanup,

      // Wrapped commander - all methods throw ActionStoppedError if stopped
      commander: wrappedCommander,

      // Original commander (use carefully)
      rawCommander: commander,

      // Trigger name for debugging
      triggerName,
    };
  }

  /**
   * Handle navigation start - stop current action first
   */
  async function onNavigationStart() {
    await stopCurrentAction();
  }

  /**
   * Handle page ready - start matching action
   */
  async function onPageReady({ url }, commander) {
    await startAction(url, commander);
  }

  /**
   * Check if an action is currently running
   */
  function isRunning() {
    return isActionRunning;
  }

  /**
   * Get current trigger name
   */
  function getCurrentTriggerName() {
    return currentTrigger?.name || null;
  }

  /**
   * Initialize - connect to navigation manager
   * @param {Object} commander - BrowserCommander instance
   */
  function initialize(commander) {
    // Stop action before navigation starts
    navigationManager.on('onBeforeNavigate', onNavigationStart);

    // Start action when page is ready
    navigationManager.on('onPageReady', (event) =>
      onPageReady(event, commander)
    );

    log.debug(() => '📋 PageTriggerManager initialized');
  }

  /**
   * Cleanup
   */
  async function destroy() {
    await stopCurrentAction();
    triggers.length = 0;
    navigationManager.off('onBeforeNavigate', onNavigationStart);
    log.debug(() => '📋 PageTriggerManager destroyed');
  }

  return {
    pageTrigger,
    stopCurrentAction,
    getCurrentTriggerName,
    isRunning,
    initialize,
    destroy,

    // Export error class and checker
    ActionStoppedError,
    isActionStoppedError,
  };
}

File: python/src/browser_commander/core/page_trigger_manager.py
"""PageTriggerManager - Manage page triggers with stoppable actions.

This module provides the page trigger system for registering URL-based
triggers with actions that can be safely stopped when navigation occurs.
"""

from __future__ import annotations

import asyncio
import re
from typing import Any, Callable

from browser_commander.core.logger import Logger
from browser_commander.core.navigation_manager import NavigationManager


class ActionStoppedError(Exception):
    """Exception raised when an action is stopped due to navigation."""

    pass


def is_action_stopped_error(error: Exception) -> bool:
    """Check if an error is an ActionStoppedError.

    Args:
        error: The exception to check

    Returns:
        True if this is an ActionStoppedError
    """
    return isinstance(error, ActionStoppedError)


def make_url_condition(pattern: str | re.Pattern | Callable) -> Callable[[str], bool]:
    """Create a URL condition from various pattern types.

    Supports:
    - Exact match: 'https://example.com/page'
    - Wildcard: '*checkout*', '/api/*', '*.json'
    - Express-style patterns: '/vacancy/:id', 'https://hh.ru/vacancy/:vacancyId'
    - RegExp: re.compile(r'/product/\\d+')
    - Custom function: lambda url: bool

    Args:
        pattern: URL pattern to match

    Returns:
        Function that checks if URL matches the pattern
    """
    # Already a function
    if callable(pattern):
        return pattern

    # Compiled regex
    if isinstance(pattern, re.Pattern):
        return lambda url: pattern.search(url) is not None

    # String pattern
    if isinstance(pattern, str):
        # Convert wildcard pattern to regex
        if "*" in pattern:
            # Escape special regex chars except *
            escaped = re.escape(pattern).replace(r"\*", ".*")
            regex = re.compile(escaped)
            return lambda url: regex.search(url) is not None

        # Check for Express-style pattern with :param
        if ":" in pattern:
            # Convert :param to regex group
            param_pattern = re.sub(r":(\w+)", r"(?P<\1>[^/]+)", pattern)
            regex = re.compile(param_pattern)
            return lambda url: regex.search(url) is not None

        # Exact match
        return lambda url: url == pattern or url.startswith(pattern)

    msg = f"Invalid pattern type: {type(pattern)}"
    raise ValueError(msg)


def all_conditions(*conditions: Callable[[str], bool]) -> Callable[[str], bool]:
    """Combine conditions with AND logic.

    Args:
        *conditions: URL condition functions

    Returns:
        Function that returns True if all conditions match
    """
    return lambda url: all(cond(url) for cond in conditions)


def any_condition(*conditions: Callable[[str], bool]) -> Callable[[str], bool]:
    """Combine conditions with OR logic.

    Args:
        *conditions: URL condition functions

    Returns:
        Function that returns True if any condition matches
    """
    return lambda url: any(cond(url) for cond in conditions)


def not_condition(condition: Callable[[str], bool]) -> Callable[[str], bool]:
    """Negate a condition.

    Args:
        condition: URL condition function

    Returns:
        Function that returns True if condition doesn't match
    """
    return lambda url: not condition(url)


class ActionContext:
    """Context passed to trigger actions for safe execution."""

    def __init__(
        self,
        url: str,
        trigger_name: str,
        abort_signal: asyncio.Event,
        commander: Any,
    ) -> None:
        """Initialize action context.

        Args:
            url: Current page URL
            trigger_name: Name of the trigger
            abort_signal: Abort signal event
            commander: Browser commander instance
        """
        self.url = url
        self.trigger_name = trigger_name
        self._abort_signal = abort_signal
        self.commander = commander
        self._cleanup_callbacks: list[Callable] = []

    def is_stopped(self) -> bool:
        """Check if action should stop."""
        return self._abort_signal.is_set()

    def check_stopped(self) -> None:
        """Raise ActionStoppedError if action should stop."""
        if self.is_stopped():
            raise ActionStoppedError(f"Action '{self.trigger_name}' was stopped")

    async def wait(self, ms: int) -> None:
        """Safe wait that respects abort signal.

        Args:
            ms: Milliseconds to wait

        Raises:
            ActionStoppedError: If action is stopped during wait
        """
        try:
            await asyncio.wait_for(
                self._abort_signal.wait(),
                timeout=ms / 1000,
            )
            # If we get here, abort was signaled
            raise ActionStoppedError(f"Action '{self.trigger_name}' was stopped")
        except asyncio.TimeoutError:
            # Timeout means wait completed normally
            pass

    async def for_each(
        self,
        items: list[Any],
        fn: Callable[[Any, int], Any],
    ) -> list[Any]:
        """Safe iteration that checks for stop between items.

        Args:
            items: Items to iterate over
            fn: Async function to call for each item (item, index)

        Returns:
            List of results

        Raises:
            ActionStoppedError: If action is stopped during iteration
        """
        results = []
        for i, item in enumerate(items):
            self.check_stopped()
            result = await fn(item, i)
            results.append(result)
        return results

    def on_cleanup(self, callback: Callable) -> None:
        """Register cleanup callback.

        Args:
            callback: Function to call during cleanup
        """
        self._cleanup_callbacks.append(callback)

    async def cleanup(self) -> None:
        """Run all cleanup callbacks."""
        for callback in self._cleanup_callbacks:
            try:
                result = callback()
                if asyncio.iscoroutine(result):
                    await result
            except Exception:
                pass  # Ignore cleanup errors


class PageTriggerManager:
    """Manage page triggers with stoppable actions."""

    def __init__(
        self,
        navigation_manager: NavigationManager,
        log: Logger,
    ) -> None:
        """Initialize PageTriggerManager.

        Args:
            navigation_manager: NavigationManager instance
            log: Logger instance
        """
        self.navigation_manager = navigation_manager
        self.log = log
        self._triggers: list[dict] = []
        self._active_actions: list[ActionContext] = []
        self._commander: Any = None
        self._background_tasks: set[asyncio.Task] = set()

    def initialize(self, commander: Any) -> None:
        """Initialize with commander reference.

        Args:
            commander: Browser commander instance
        """
        self._commander = commander

        # Subscribe to URL changes
        self.navigation_manager.on("on_url_change", self._on_url_change)

    def _on_url_change(self, event: dict) -> None:
        """Handle URL change event."""
        new_url = event["new_url"]

        # Stop all active actions
        for ctx in self._active_actions:
            ctx._abort_signal.set()

        # Check triggers for new URL
        task = asyncio.create_task(self._check_triggers(new_url))
        self._background_tasks.add(task)
        task.add_done_callback(self._background_tasks.discard)

    async def _check_triggers(self, url: str) -> None:
        """Check and run matching triggers for URL.

        Args:
            url: Current URL to check
        """
        for trigger in self._triggers:
            condition = trigger["condition"]
            if condition(url):
                await self._run_trigger(trigger, url)

    async def _run_trigger(self, trigger: dict, url: str) -> None:
        """Run a trigger's action.

        Args:
            trigger: Trigger configuration
            url: Current URL
        """
        name = trigger.get("name", "unnamed")
        action = trigger["action"]

        self.log.debug(lambda: f"Running trigger '{name}' for URL: {url}")

        abort_signal = asyncio.Event()
        ctx = ActionContext(
            url=url,
            trigger_name=name,
            abort_signal=abort_signal,
            commander=self._commander,
        )
        self._active_actions.append(ctx)

        try:
            await action(ctx)
        except ActionStoppedError:
            self.log.debug(lambda: f"Trigger '{name}' was stopped")
        except Exception as e:
            self.log.debug(lambda _e=e: f"Trigger '{name}' error: {_e}")
        finally:
            await ctx.cleanup()
            if ctx in self._active_actions:
                self._active_actions.remove(ctx)

    def page_trigger(self, config: dict) -> None:
        """Register a page trigger.

        Args:
            config: Trigger configuration with:
                - condition: URL condition (string, regex, or function)
                - action: Async function to run when condition matches
                - name: Optional trigger name for debugging
        """
        condition = config.get("condition")
        action = config.get("action")
        name = config.get("name", "unnamed")

        if not condition or not action:
            msg = "page_trigger requires 'condition' and 'action'"
            raise ValueError(msg)

        # Convert condition to callable
        if not callable(condition):
            condition = make_url_condition(condition)

        self._triggers.append(
            {
                "condition": condition,
                "action": action,
                "name": name,
            }
        )

    async def destroy(self) -> None:
        """Clean up and stop all actions."""
        # Stop all active actions
        for ctx in self._active_actions:
            ctx._abort_signal.set()
            await ctx.cleanup()

        self._active_actions.clear()
        self._triggers.clear()

        # Unsubscribe from events
        self.navigation_manager.off("on_url_change", self._on_url_change)
